libcxx/include/__flat_set/flat_set.h

```diff
@@ -0,0 +1,856 @@
+  _LIBCPP_HIDE_FROM_ABI flat_set(__ctor_uses_allocator_empty_tag, const _Allocator& __alloc, _CompArg&&... __comp)
+      : __keys_(std::make_obj_using_allocator<container_type>(__alloc)),
+        __compare_(std::forward<_CompArg>(__comp)...) {}
+
+  _LIBCPP_HIDE_FROM_ABI bool __is_sorted_and_unique(auto&& __key_container) const {
+    auto __greater_or_equal_to = [this](const auto& __x, const auto& __y) { return !__compare_(__x, __y); };
+    return ranges::adjacent_find(__key_container, __greater_or_equal_to) == ranges::end(__key_container);
+  }
+
+  // This function is only used in constructors. So there is not exception handling in this function.
+  // If the function exits via an exception, there will be no flat_set object constructed, thus, there
+  // is no invariant state to preserve
+  _LIBCPP_HIDE_FROM_ABI void __sort_and_unique() {
+    ranges::sort(__keys_, __compare_);
+    auto __dup_start = ranges::unique(__keys_, __key_equiv(__compare_)).begin();
+    __keys_.erase(__dup_start, __keys_.end());
+  }
+
+  template <class _InputIterator>
+  _LIBCPP_HIDE_FROM_ABI void __append(_InputIterator __first, _InputIterator __last) {
+    __keys_.insert(__keys_.end(), std::move(__first), std::move(__last));
+  }
+
+  template <class _Range>
+  _LIBCPP_HIDE_FROM_ABI void __append(_Range&& __rng) {
+    if constexpr (requires { __keys_.insert_range(__keys_.end(), std::forward<_Range>(__rng)); }) {
+      // C++23 Sequence Container should have insert_range member function
+      __keys_.insert_range(__keys_.end(), std::forward<_Range>(__rng));
+    } else if constexpr (ranges::common_range<_Range>) {
+      __keys_.insert(__keys_.end(), ranges::begin(__rng), ranges::end(__rng));
+    } else {
+      for (auto&& __x : __rng) {
+        __keys_.insert(__keys_.end(), std::forward<decltype(__x)>(__x));
+      }
+    }
+  }
+
+  template <bool _WasSorted, class... _Args>
+  _LIBCPP_HIDE_FROM_ABI void __append_sort_merge_unique(_Args&&... __args) {
+    auto __on_failure    = std::__make_exception_guard([&]() noexcept { clear() /* noexcept */; });
+    size_type __old_size = size();
+    __append(std::forward<_Args>(__args)...);
+    if (size() != __old_size) {
+      if constexpr (!_WasSorted) {
+        ranges::sort(__keys_.begin() + __old_size, __keys_.end(), __compare_);
+      } else {
+        _LIBCPP_ASSERT_SEMANTIC_REQUIREMENT(__is_sorted_and_unique(__keys_ | ranges::views::drop(__old_size)),
+                                            "Either the key container is not sorted or it contains duplicates");
+      }
+      ranges::inplace_merge(__keys_.begin(), __keys_.begin() + __old_size, __keys_.end(), __compare_);
+
+      auto __dup_start = ranges::unique(__keys_, __key_equiv(__compare_)).begin();
+      __keys_.erase(__dup_start, __keys_.end());
+    }
+    __on_failure.__complete();
+  }
+
+  template <class _Self, class _Kp>
+  _LIBCPP_HIDE_FROM_ABI static auto __find_impl(_Self&& __self, const _Kp& __key) {
```

Comment:
You might be able to remove this if the `iterator` and the `const_iterator` have the same type?